### ISSUE TITLE
BL Prod Support: Refactor exporter issue and speed issues

### ISSUE THREAD

**Issue by ShanaLMoore** [scientist-softserv/britishlibrary#289]
## Problem Statement

The exporter, as implemented, has in-efficient memory and spends considerable time in a loop of queries, which are non-performant.

The goal is to remove the memory in-efficiencies so that exports can work:

- Remove calls to ActiveFedora::Base, favor SOLR query (which will require some research)
- Reduce the "expand/contract/expand" pattern of variables and testing values.
- When exporting (and perhaps importing) on a property that is declared as an object; only export as if object if the entry's property is an object (instead of a value).
- Memoization of the  `#current_record_ids` as those could be getting rebuilt each time we call the method.
- Refactor create new entries

```ruby
def something
  @something ||= do_the_expensive_thing
end
```

## Conversations

These are the two primary Slack conversations to provide some background:

- https://assaydepot.slack.com/archives/C0313NK2LJ0/p1676635112118209
- https://assaydepot.slack.com/archives/C0313NK2LJ0/p1676886280688969

### Raw Notes (from Rob)

https://gist.github.com/orangewolf/9819d778264d0c836bb716428e92d021

### Raw Notes (from Jeremy)

Look into ApplicationParser#create_new_entries; refactor the Exporter.export to avoid the indirection

- ApplicationParser#current_record_ids is confusing
- find_child_file_sets calls Fedora directly instead with a big heavy lift.  This is thing one that needs to happen.
- We don’t memoize the underlying current_record_ids
- ApplicationParser#create_new_entries; take a set of ids and an entry class
- The find_or_create doesn’t check the status, we run all of the work jobs inline; the perform needs to be now because the export file writes the file, thus the job needs everything to run (perhaps unsolveable)
- Bulkrax::ExportWorkJob, entry.build regardless of FileSet or Work uses the same logic, but doesn’t handle the creator correctly (creator is complex object on work type; on file set it is not complex).  The definition of creator in the parser causes problems. build_metadata_mapping needs to attempt build_object if the parser says its an object and the factory_class says the property is an object

**Comment by jeremyf** [comment]
We start from  [Bulkrax::ExporterJob#perform](https://github.com/samvera-labs/bulkrax/blob/af91434e8e628c129b0e35aa71aa43dd309cd76a/app/jobs/bulkrax/exporter_job.rb#L3-L15)

```ruby
module Bulkrax
  class ExporterJob < ApplicationJob
    queue_as :export

    def perform(exporter_id)
      exporter = Exporter.find(exporter_id)
      exporter.export
      exporter.write
      exporter.save
      true
    end
  end
end
```

And right away dive into the [Bulkrax::Exporter#export](https://github.com/samvera-labs/bulkrax/blob/af91434e8e628c129b0e35aa71aa43dd309cd76a/app/models/bulkrax/exporter.rb#L19-L33) method.

```ruby
def export
  current_run && setup_export_path
  case self.export_from
  when 'collection'
    create_from_collection
  when 'importer'
    create_from_importer
  when 'worktype'
    create_from_worktype
  when 'all'
    create_from_all
  end
rescue StandardError => e
  status_info(e)
end
```

A quick win refactor is to shift from the case statement to a `send` approach.

Then we want to look at [Bulkrax::CsvParser#create\_new\_entries](https://github.com/samvera-labs/bulkrax/blob/af91434e8e628c129b0e35aa71aa43dd309cd76a/app/parsers/bulkrax/csv_parser.rb#L253-L278)

```ruby
def create_new_entries
  current_record_ids.each_with_index do |id, index|
    break if limit_reached?(limit, index)

    this_entry_class = if @collection_ids.include?(id)
                         collection_entry_class
                       elsif @file_set_ids.include?(id)
                         file_set_entry_class
                       else
                         entry_class
                       end
    new_entry = find_or_create_entry(this_entry_class, id, 'Bulkrax::Exporter')

    begin
      entry = ExportWorkJob.perform_now(new_entry.id, current_run.id)
    rescue => e
      Rails.logger.info("#{e.message} was detected during export")
    end

    self.headers |= entry.parsed_metadata.keys if entry
  end
end
alias create_from_collection create_new_entries
alias create_from_importer create_new_entries
alias create_from_worktype create_new_entries
alias create_from_all create_new_entries
```

Here we have the create\_new\_entries and the various methods that are aliases.  We want to look into [Bulkrax::CsvParser#current\_record\_ids](https://github.com/samvera-labs/bulkrax/blob/af91434e8e628c129b0e35aa71aa43dd309cd76a/app/parsers/bulkrax/csv_parser.rb#L188-L215)

```ruby
def current_record_ids
  @work_ids = []
  @collection_ids = []
  @file_set_ids = []

  case importerexporter.export_from
  when 'all'
    @work_ids = ActiveFedora::SolrService.query("has_model_ssim:(#{Hyrax.config.curation_concerns.join(' OR ')}) #{extra_filters}", method: :post, rows: 2_147_483_647).map(&:id)
    @collection_ids = ActiveFedora::SolrService.query("has_model_ssim:Collection #{extra_filters}", method: :post, rows: 2_147_483_647).map(&:id)
    @file_set_ids = ActiveFedora::SolrService.query("has_model_ssim:FileSet #{extra_filters}", method: :post, rows: 2_147_483_647).map(&:id)
  when 'collection'
    @work_ids = ActiveFedora::SolrService.query(
      "member_of_collection_ids_ssim:#{importerexporter.export_source + extra_filters} AND has_model_ssim:(#{Hyrax.config.curation_concerns.join(' OR ')})", method: :post, rows: 2_000_000_000
    ).map(&:id)
    # get the parent collection and child collections
    @collection_ids = ActiveFedora::SolrService.query("id:#{importerexporter.export_source} #{extra_filters}", method: :post, rows: 2_147_483_647).map(&:id)
    @collection_ids += ActiveFedora::SolrService.query("has_model_ssim:Collection AND member_of_collection_ids_ssim:#{importerexporter.export_source}", method: :post,
                                                       rows: 2_147_483_647).map(&:id)
    find_child_file_sets(@work_ids)
  when 'worktype'
    @work_ids = ActiveFedora::SolrService.query("has_model_ssim:#{importerexporter.export_source + extra_filters}", method: :post, rows: 2_000_000_000).map(&:id)
    find_child_file_sets(@work_ids)
  when 'importer'
    set_ids_for_exporting_from_importer
  end

  @work_ids + @collection_ids + @file_set_ids
end
```

This logic is creating several instance variables.  We want to consider what we’re passing back and setting.  Most importantly we want to look at  [Bulkrax::CsvParser#find\_child\_file\_sets](https://github.com/samvera-labs/bulkrax/blob/af91434e8e628c129b0e35aa71aa43dd309cd76a/app/parsers/bulkrax/csv_parser.rb#L219-L223).

```ruby
def find_child_file_sets(work_ids)
  work_ids.each do |id|
    ActiveFedora::Base.find(id).file_set_ids.each { |fs_id| @file_set_ids << fs_id }
  end
end
```

**Comment by AKFosterDev** [comment]
Tested with this csv that include trailing commas
[sw_bl_format.csv](https://github.com/scientist-softserv/britishlibrary/files/10855715/sw_bl_format.csv)

**Comment by jeremyf** [comment]
This skipped verification because the logic is now in production.  Why is this in production?  To work through the outage, we opted to deploy the main branch to production.

The following commits are what made it into production:


- f443510f — add init argument so that the isni is judiciciously managed Rory McNicholl, (2023-02-27) (origin/stop_removing_isni_when_loading_form)
- 832de4e3 — update bulkrax ref Amanda, (2023-02-28) (origin/i289-bulkrax-upgrade)
- 4cf909bc — upgrade bulkrax Amanda, (2023-02-27)
- b7d7b5d2 — Addressing exporter memory issue in Bulkrax Jeremy Friesen, (2023-02-22) (i289-jeremyf)
- ee88c7db — Hot Fix: "undefined local variable or method `section`" Jeremy Friesen, (2023-02-17) (i284)
- c8472e9f — Add xml fixture for future testing Kirk Wang, (2023-02-15)
- ab56f6f1 — Add yarn install towards the end of build Kirk Wang, (2023-02-14)
- 2956c480 — Change header to current_he_institution_name Kirk Wang, (2023-02-14)
- 27097558 — Adding spec that demonstrates the bug Jeremy Friesen, (2023-02-10) (adding-spec-to-prove-bug)
- af5c8584 — upgrade bulrkax to 5+ Shana, (2023-02-15)

**Comment by grahamjevon** [comment]
Issues identified while testing in production.

Test: 

2. Created [New Exporter](https://bl-demo.iro.bl.uk/exporters/45?locale=en) on bl-demo.iro.bl.uk
3. Type of Export: Metadata Only
4. Export From: Work Type
5. Work Type: Article
6. Export Format: CSV

Result:

When I downloaded the export, it had only 1 row of data in the csv. However, bl-demo.iro.bl.uk contains 492 articles (according to the facet on the explore the research repository page). I would therefore expect 492 rows of data.

NB: The exporter summary page suggests there were 494 total results (on the Work Entries" tab) - all of which appear to be "Complete" (though I didn't look at every page of the results). Meanwhile, the "File Set Entries" tab has 159 results - all of which appear to have "Failed".

I repeated the test for the work type books and received the same unexpected result: an exported csv that contains only 1 row of data.

**Comment by AKFosterDev** [comment]
## QA Instructions

Test: on Staging

1. Create [New Exporter](https://bl.bl-staging.notch8.cloud/exporters?locale=en) on bl-staging
2. Type of Export: Metadata Only
3. Export From: Work Type
4. Work Type: Article
5. Export Format: CSV

**Comment by ShanaLMoore** [comment]
# QA RESULTS:

**Comment by grahamjevon** [comment]
@ShanaLMoore , while the export now successfully exports multiple rows, it also seems to generate rogue data. When I downloaded your test exporter it had 46 rows of "Article" Work Type data, as expected. But it also contained one additional row of data from the "GenericWork" Work Type, which was not expected. I replicated this error when I ran an export with the same parameters.


![Image](https://user-images.githubusercontent.com/51366817/224962564-6e5d7730-3beb-4b99-9204-e88565fafdc3.png)

I then ran a test exporter for the "GenericWork" work type. Similarly, this generated 10 rows of GenericWork data, as expected. But it also included two rows of "FileSet" data, which was not expected.


![Image](https://user-images.githubusercontent.com/51366817/224963974-08117056-5e6a-4b83-a8c9-dc9ca3e39cca.png)


![Image](https://user-images.githubusercontent.com/51366817/224963442-49e152ed-e43d-4c22-84e2-c56e4e1add6d.png)

This looks like it might be related to the issue that @jeremyf filed: https://github.com/samvera-labs/bulkrax/issues/754. Because in both these tests, the Work Entries tab on the exporter summary had the correct number of entries. And in both instances, the erroneous data is listed in the FileSet entries tab. The one erroneous row in the csv for the "Article" work type is listed in the "FileSet" tab.

![Image](https://user-images.githubusercontent.com/51366817/224964986-ee0f9fb2-900b-476e-b217-317f4bdc06be.png)

Meanwhile, for the "GenericWork" test, the two erroneous csv rows are also listed in the "FileSet" tab (highlighted yellow). Though, here there is a 3rd item in the "FileSet" tab (the middle row, entry id 616), which seems to be the same as entry id 605 in the "Work Entries" tab.


![Image](https://user-images.githubusercontent.com/51366817/224967408-78b13040-c30e-4c37-aea9-168ee7b84350.png)



![Image](https://user-images.githubusercontent.com/51366817/224967498-e46a762c-6c9a-46c8-95fb-ae1e93fee86c.png)

**Comment by ShanaLMoore** [comment]
@grahamjevon thank you for the thorough review. I must admit that I totally missed out on the details of the export, and was mainly focused on it no longer producing 1 row 🙃 🙈 

I'll have @jeremyf review this and likely pull it back until the issue you mentioned is resolved, too.

**Comment by jeremyf** [comment]
**tl;dr**: The export of a work type should be exporting the works's file sets.  The underlying problem is not a problem with the Bulkrax upgrade but is instead some other data assignment problem; perhaps something to do with PDF splitting.  

This is an approximation of the query that runs to get the Works and then their "File Sets":

`ActiveFedora::SolrService.query("has_model_ssim:GenericWork ", { fl: "id,file_set_ids_ssim,has_model_ssim", method: :post, rows: 10_000 })`

(I include the `has_model_ssim` to see if we're getting a mix of results)

The results are as follows:

```ruby
works = [
  {
    "has_model_ssim"=>["GenericWork"],
    "id"=>"7d99a819-9e71-41c3-a3a9-8ab147feb791"
  }, {
    "has_model_ssim"=>["GenericWork"],
    "id"=>"9918069a-5665-433d-a0c2-94586c398eb0",
    "file_set_ids_ssim"=>["e1cc5173-60b5-4f39-8d8d-d9f78e7fb1ef", "7d99a819-9e71-41c3-a3a9-8ab147feb791"]
  }, {
    "has_model_ssim"=>["GenericWork"],
    "id"=>"177b102e-9e3b-47b5-837e-f7d02e43b9fa"
  }, {
    "has_model_ssim"=>["GenericWork"],
    "id"=>"93866f95-745e-46ce-b1e0-84bd673a1528"
  }, {
    "has_model_ssim"=>["GenericWork"],
    "id"=>"671811e2-7dd8-4cf6-99fe-b0b52af972c1"
  }, {
    "has_model_ssim"=>["GenericWork"],
    "id"=>"2d93ebff-f83c-4aa9-af4f-b1fd6ac1ad39"
  }, {
    "has_model_ssim"=>["GenericWork"],
    "id"=>"9491a6f3-15e2-4bd2-84f1-a239e7269554"
  }, {
    "has_model_ssim"=>["GenericWork"],
    "id"=>"53ae3a7c-7569-43d3-835c-a27a2ae13c17",
    "file_set_ids_ssim"=>["26b6ba1c-44ee-4718-bc5c-864fe3503f01"]
  }, {
    "has_model_ssim"=>["GenericWork"],
    "id"=>"baeffbd5-0a4b-4cd2-a3b6-c22503e42f6c"
  }, {
    "has_model_ssim"=>["GenericWork"],
    "id"=>"aa7c40a1-fbdd-445f-a091-aa935ce2a5c3"
  }
]
```

Using the `works` variable that I set, I then find all of the Fedora records for the `file_set_ids_ssim`

```ruby
file_sets = works.map do |w|
  w['file_set_ids_ssim'] || []
end.flatten.compact.map do |fs_id|
  ActiveFedora::Base.find(fs_id)
end
```

And then ask what are all of the classes of the file\_sets (e.g. `file_sets.map { |fs| { id: fs.id, class: fs.class } }`)

```ruby
[
  {
    id: "e1cc5173-60b5-4f39-8d8d-d9f78e7fb1ef",
    class: FileSet
  }, {
    id: "7d99a819-9e71-41c3-a3a9-8ab147feb791",
    class: GenericWork
  }, {
    id: "26b6ba1c-44ee-4718-bc5c-864fe3503f01",
    class: FileSet
  }
]
```

**Comment by jeremyf** [comment]
In IIIF Print we push the child works into the `file_set_ids_ssim`; this is so that we can render in UniversalViewer a parent PDF's split JPG pages.

-   **Option 1:** We don't put child works in the `file_set_ids_ssim` field.
-   **Option 2:** We query the `file_set_ids_ssim` to only select `has_model_ssim:FileSet`.

There are other options

**Comment by jeremyf** [comment]
Further diving into this, and this is in fact a bug introduced into Bulkrax.

**Comment by AKFosterDev** [comment]
## SoftServ QA Passes


### Article Work Type

Test: on Staging

1. Create [New Exporter](https://bl.bl-staging.notch8.cloud/exporters/18?locale=en) on bl-staging
2. Type of Export: Metadata Only
3. Export From: Work Type
4. Work Type: Article
5. Export Format: CSV
6. Create and export

From exporters list:
- [x] From the newly created export click download button
- [x] CSV should should populate multiple rows
- [x] All rows show 'Article' work type

![Image](https://user-images.githubusercontent.com/39319859/225698827-f297673d-d81c-438b-889c-4916f90ae4c6.png)

### Generic Work Type

Test: on Staging

1. Create [New Exporter](https://bl.bl-staging.notch8.cloud/exporters/19?locale=en) on bl-staging
2. Type of Export: Metadata Only
3. Export From: Work Type
4. Work Type: Generic Work
5. Export Format: CSV
6. Create and export

From exporters list:
- [x] From the newly created export click download button
- [x] CSV should should populate multiple rows
- [x] 10 rows show 'GenericWork' work type and 2 rows show 'FileSet' (no double sha's)

**10 Work Entries Created**

![Image](https://user-images.githubusercontent.com/39319859/225699335-292a6fbd-46aa-474b-863e-5217151947cd.png)

**2 File Sets Created**

![Image](https://user-images.githubusercontent.com/39319859/225699476-4bb7bef5-59bd-4100-b01a-bcca628b1e00.png)

**Downloaded CSV**


![Image](https://user-images.githubusercontent.com/39319859/225699670-e1023827-f643-4ecd-929f-8271cd02f1a9.png)

**Comment by grahamjevon** [comment]
My initial tests on staging worked as expected on Friday. But when I came to continue testing today (Monday morning), staging is empty of works. I therefore added some works today (one via the UI and 3 via BX) so that I could continue testing. But no subsequent exporters have been successful. 

**[Failed Test example](https://bl.bl-staging.notch8.cloud/exporters/28?locale=en)**

![Image](https://user-images.githubusercontent.com/51366817/226307880-d81c2543-b282-4f83-a80e-5ec7f52da50b.png)


![Image](https://user-images.githubusercontent.com/51366817/226309810-2bcce5e6-b483-4642-bc2b-9bbabfc0bc8d.png)


Export of articles has failed. Albeit, I can download a csv (though it only contains 1 row of data when it should return 2. 

The next screenshot shows 2 articles on the front end search results

![Image](https://user-images.githubusercontent.com/51366817/226308872-134e8f63-a33d-444e-aafa-38ec0fc8655a.png)

The exporter has a solr error.

**Comment by jeremyf** [comment]
> Staging was not properly storing files at all before. 

https://assaydepot.slack.com/archives/C0313NK2LJ0/p1679523808125029

I'm wondering how this relates.  I'd like to get this stabilized before digging further in.

**Comment by ShanaLMoore** [comment]
update: the staging deploy issues have been resolved

**Comment by ShanaLMoore** [comment]
On a new tenant, I created this [importer](https://qa.bl-staging.notch8.cloud/importers?locale=en) 

I imported the following CSV (I since needed to add an institution to the parent work so that the UV button would render): https://qa.bl-staging.notch8.cloud/importers?locale=en

I then created a metadata only [exporter](https://qa.bl-staging.notch8.cloud/exporters?locale=en), work type: Image. It generated the following csv: [export_Image_from_worktype_1.csv](https://github.com/scientist-softserv/britishlibrary/files/11081805/export_Image_from_worktype_1.csv)

**Comment by ShanaLMoore** [comment]
@grahamjevon I believe this is stuck in pending because of what looks to be a bottleneck happening in sidekiq. The exporter also depends on sidekiq. Its job will get in the back of the queue of whatever processes are happening. 



![Image](https://user-images.githubusercontent.com/10081604/228271421-5b9f23da-6619-427b-bd07-697d6e42342a.png)



We will be looking into all of it today.

**Comment by ShanaLMoore** [comment]
@grahamjevon Could you attach the files you uploaded for the two importers you created, to this thread?

**Comment by NoraRamsey** [comment]
[BX Import template.csv](https://github.com/scientist-softserv/britishlibrary/files/11092572/BX.Import.template.csv)


@ShanaLMoore Here is the csv I imported!

**Comment by ShanaLMoore** [comment]
Blocked until we've reconcile prod with staging deploys. ref #376

**Comment by ShanaLMoore** [comment]
unblocked, if tested in staging.

**Comment by grahamjevon** [comment]
Hi @ShanaLMoore , should the exporter expected to be functioning in staging when selecting "Export From: All"? Currently, when I select these parameters and click on "Create and Export" literally nothing happens. (NB: It does appear to function when exporting by work type - though I haven't tested this thoroughly yet).

![Image](https://user-images.githubusercontent.com/51366817/232808102-3f465328-e67b-477c-845a-ecb2fac4a239.png)

**Comment by ShanaLMoore** [comment]
Kirk [updated bulkrax](https://github.com/scientist-softserv/britishlibrary/pull/389), which pulls in a fix that should unblock the non-functioning button [Graham described](https://github.com/scientist-softserv/britishlibrary/issues/289#issuecomment-1513257407).

**Comment by ShanaLMoore** [comment]
this was a prod support issue. This code is already deployed to prod so QA isn't required. This can be closed if the exporter button works again.  cc @grahamjevon

**Comment by kirkkwang** [comment]
tested that the Export and Create button does work and created this:
https://mola.bl-staging.notch8.cloud/exporters/22?locale=en

signifying that the issue is no longer happening, closing ticket
